src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/catalog.cpp

```diff
@@ -81,72 +82,373 @@ struct component
+    this_component->module_name = fileName;
+    HRESULT hr = xmlReader->MoveToFirstAttribute();
+    // Using this pattern intead of calling multiple MoveToAttributeByName improves performance
+    const WCHAR* activatableClass = nullptr;
+    const WCHAR* threadingModel = nullptr;
+    const WCHAR* xmlns = nullptr;
+    if (S_FALSE == hr)
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+    else
+    {
+        while (TRUE)
+        {
+            const WCHAR* pwszLocalName;
+            const WCHAR* pwszValue;
+            if (FAILED(hr = xmlReader->GetLocalName(&pwszLocalName, NULL)))
+            {
+                wprintf(L"Error getting local name, error is %08.8lx", hr);
+                return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+            }
+            if (FAILED(hr = xmlReader->GetValue(&pwszValue, NULL)))
+            {
+                wprintf(L"Error getting value, error is %08.8lx", hr);
+                return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+            }
+            if (_wcsicmp_l(L"threadingModel", pwszLocalName, locale) == 0)
+            {
+                threadingModel = pwszValue;
+            }
+            else if (_wcsicmp_l(L"name", pwszLocalName, locale) == 0)
+            {
+                activatableClass = pwszValue;
+            }
+            else if (_wcsicmp_l(L"xmlns", pwszLocalName, locale) == 0)
+            {
+                xmlns = pwszValue;
+            }
+
+            if (xmlReader->MoveToNextAttribute() != S_OK)
+            {
+                break;
             }
         }
     }
+    if (threadingModel == nullptr)
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+    if (_wcsicmp_l(L"sta", threadingModel, locale) == 0)
+    {
+        this_component->threading_model = ABI::Windows::Foundation::ThreadingType::ThreadingType_STA;
+    }
+    else if (_wcsicmp_l(L"mta", threadingModel, locale) == 0)
+    {
+        this_component->threading_model = ABI::Windows::Foundation::ThreadingType::ThreadingType_MTA;
+    }
+    else if (_wcsicmp_l(L"both", threadingModel, locale) == 0)
+    {
+        this_component->threading_model = ABI::Windows::Foundation::ThreadingType::ThreadingType_BOTH;
+    }
+    else
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+
+    if (activatableClass == nullptr || !activatableClass[0])
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+    this_component->xmlns = xmlns; // Should we care if this value is blank or missing?
+    // Check for duplicate activatable classes
+    auto component_iter = g_types.find(activatableClass);
+    if (component_iter != g_types.end()) 
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_DUPLICATE_ACTIVATABLE_CLASS);
+    }
+    g_types[activatableClass] = this_component;
+    return S_OK;
+}
 
+HRESULT ParseDependentAssemblyTag(ComPtr<IXmlReader> xmlReader)
```

Comment:
Should these methods be void and use throwing instead of "return error" ? Or should there be a try/catch at the top level since you're mixing C++ constructs and non-throwing constructs?

Reply:
I am returning errors such as ERROR_SXS_MANIFEST_PARSE_ERROR and ERROR_SXS_DUPLICATE_ACTIVATABLE_CLASS so I don't really want it to be void. Maybe just a try catch at the top level.

Reply:
Should I go with this solution?

    try
    {
        return ParseRootManifestFromXmlReaderInput(fileStream.Get());
    }
    catch(...)
    {
        LOG_CAUGHT_EXCEPTION();
        return ERROR_SXS_MANIFEST_PARSE_ERROR;
    }